dora_node_hub/dora_gemini_diff_drive_navigation/dora_gemini_diff_drive_navigation/main.py

```diff
@@ -0,0 +1,144 @@
+"""Main entry point for the dora_gemini_diff_drive_navigation package."""
+
+import asyncio
+import os
+import threading
+
+import cv2 as cv
+import numpy as np
+import pyarrow as pa
+from dora import Node
+
+from dora_gemini_diff_drive_navigation.controller import DiffDriveGeminiControl
+
+
+def start_event_loop(loop: asyncio.AbstractEventLoop) -> None:
+    """Starts the asyncio event loop in the background thread."""
+    asyncio.set_event_loop(loop)
+    loop.run_forever()
+
+
+def zero_twist() -> np.ndarray:
+    """Create a zero twist array."""
+    return np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0], dtype=np.float64)
+
+
+def main() -> None:
+    """TODO: Add docstring."""
+    node = Node()
+
+    # Required node configuration
+    command = os.getenv("COMMAND", "")
+    if not command:
+        print("COMMAND environment variable is not set. Using command via input event.")
+    # Optional node configuration
+    model = os.getenv("MODEL", "gemini-2.5-flash")
+
+    # Initialize the controller with the specified model.
+    controller = DiffDriveGeminiControl(model=model)
+    last_image = None
+    # Initialize cmd_vel with zero velocity.
+    cmd_vel = zero_twist()
+
+    # Setup for the background event loop thread
+    loop = asyncio.new_event_loop()
+    thread = threading.Thread(target=start_event_loop, args=(loop,), daemon=True)
+    thread.start()
+
+    generated_velocities_future = None
+
+    for event in node:
+        event_type = event["type"]
+        if event_type == "INPUT":
+            event_id = event["id"]
+            if event_id == "image":
+                storage = event["value"]
+                metadata = event["metadata"]
+                encoding = metadata["encoding"]
+                width = metadata["width"]
+                height = metadata["height"]
+
+                if encoding in {"bgr8", "rgb8"}:
+                    channels = 3
+                    storage_type = np.uint8
+                else:
+                    raise RuntimeError(f"Unsupported image encoding: {encoding}")
+
+                frame = storage.to_numpy().astype(storage_type).reshape((height, width, channels))
+                if encoding == "bgr8":
+                    pass
+                elif encoding == "rgb8":
+                    frame = frame[:, :, ::-1]  # Convert RGB to BGR
+                else:
+                    raise RuntimeError(f"Unsupported image encoding: {encoding}")
+
+                # Convert the frame to png
+                ret, frame = cv.imencode(".png", frame)
+                if not ret:
+                    raise RuntimeError("Failed to encode image to PNG format.")
+                last_image = frame
```

Comment:
@francocipollone meta: It may be worthwhile to isolate this into a function that can fetch an OpenCV image from an image event. I'm surprised it doesn't exist already.

Reply:
Done